src/System.Runtime.Extensions/src/System/IO/StreamWriter.cs

```diff
@@ -1023,9 +1299,66 @@ private bool HaveWrittenPreamble_Prop
             // Services guys have some perf tests where flushing needlessly hurts.
             if (flushStream)
             {
+                _this.ReturnBuffers();
                 await stream.FlushAsync().ConfigureAwait(false);
             }
         }
+
+        private static async Task FlushAsyncInternalAwaitWrite(Task writeTask, StreamWriter _this, Stream stream)
+        {
+            await writeTask.ConfigureAwait(false);
+
+            _this.ReturnBuffers();
+            await stream.FlushAsync().ConfigureAwait(false);
+        }
+
+        private static Task WritePreambleAsync(StreamWriter _this, Encoding encoding, Stream stream)
+        {
+            _this.HaveWrittenPreamble_Prop = true;
+            byte[] preamble = encoding.GetPreamble();
+            if (preamble.Length > 0)
+            {
+                return stream.WriteAsync(preamble, 0, preamble.Length);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static int EncodeStringBytes(string value, int offset, int length, byte[] byteBuffer, Encoder encoder, bool flushEncoder)
+            => encoder.GetBytes(value.AsReadOnlySpan().Slice(offset, length), new Span<byte>(byteBuffer), flushEncoder);
+
         #endregion
+
+        private char[] CharBuffer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _rentedCharBuffer ?? RentCharBuffer();
+        }
+
+        private byte[] ByteBuffer
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _rentedByteBuffer ?? RentByteBuffer();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private char[] RentCharBuffer() => _rentedCharBuffer = ArrayPool<char>.Shared.Rent(_charLen);
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private byte[] RentByteBuffer() => _rentedByteBuffer = ArrayPool<byte>.Shared.Rent(_byteLen);
+
+        private void ReturnBuffers()
+        {
+            byte[] byteBuffer = Interlocked.Exchange(ref _rentedByteBuffer, null);
```

Comment:
Accidentally returning buffers that might still get used makes me nervous. Are there any tests / asserts we can add to help validate / prevent this sort of issue?

Reply:
Discussed here:
https://github.com/dotnet/corefx/pull/23727#discussion_r136565768

The short of it is, we could change the code such that whenever it gets the buffer to use, it nulls it out, and then puts it back when it's done.  That can help to minimize the impact of misuse.  Then there's the question of whether those operations are interlocked or not: interlocking would help to guarantee correctness in the face of multithreaded misuse, but at a non-trivial expense that would be paid even in the correctly-used case.

Reply:
Hmm. So always rent a buffer for each method that needs it? It would be interesting to see what that does to perf.

Reply:
> So always rent a buffer for each method that needs it?

I'm saying rather than doing:
```C#
char[] buffer = _rentedCharBuffer ?? RentBuffer();
DoOperation(buffer);
```
instead doing:
```C#
char[] buffer = _rentedCharBuffer;
if (buffer != null)
{
    _rentedCharBuffer = null;
}
else
{
    buffer = ArrayPool<char>.Shared.Rent(_charLen);
}

DoOperation(buffer);

if (_rentedCharBuffer != null) throw ...;
_rentedCharBuffer = buffer;
```
or with interlockeds:
```C#
char[] buffer = Interlocked.Exchange(ref _rentedCharBuffer, null) ??
 ArrayPool<char>.Shared.Rent(_charLen);

DoOperation(buffer);

if (Interlocked.Exchange(ref _rentedCharBuffer, buffer) != null) throw ...;
```
The latter would add non-trivial cost.  I wouldn't expect the former to be particularly measurable, though I could be wrong.

Reply:
So just a best effort to flag using the class on multiple threads? Would introducing a throw be appropriate from a compat perspective?

Reply:
> So just a best effort 

Right.

> Would introducing a throw be appropriate from a compat perspective?

I think it would, though I'd be interested in others opinions.  As it currently stands concurrently writing on multiple threads can already result in a multitude of exceptions.  Of course we wouldn't have to throw.

Reply:
At a quick look we are also paying the cost of the two interlockeds whenever flushStream is true. That said I wouldn't expect code calling `Flush` at any noticeable rate to care that much about performance. Anyway it seems worth to look if there are no other common patterns, that may trigger these extra interlockeds for which, perf is important... I am almost saying to drop this best effort... :smile: please, convince me otherwise :smile:

Reply:
The `Exchange` is to prevent a double add to array pool which would have non-localized side effects; as the same array could then be concurrently rented by well behaved code.

However things would already be in a bad way locally by performing parallel unlocked Writes to the StreamWriter... 🤷‍♂️

Reply:
The path to hit multiple calls to `ReturnBuffers` is calling flush explicitly or having `AutoFlush` set to true, so for me the current two interlocked calls in `ReturnBuffers` seem like a good trade off.

Regarding the best effort to throw a different exception I am not sure if it is worth the trouble (albeit small) since we can't really control how the issue will manifest in each run and we can't really prevent data corruption. Of course in a large number of runs with such bugs we should hit it some times but my impression is that the other "random" errors (likely null ref in this case and data corruption on the stream) already provide a pretty good indication of the issue...*but* I am fine with it if you opt to add it...